Source/Fuse.Nodes/Visual.Children.SafeIterator.uno

```diff
@@ -0,0 +1,171 @@
+using Uno;
+using Uno.Collections;
+using Uno.Text;
+using Uno.UX;
+
+using Fuse.Internal;
+
+namespace Fuse
+{
+	public partial class Visual
+	{
+		// Performance critical code paths should not need to create this array.
+		// This is here to optimize the corner cases where we need indexed lookup.
+		// The SafeIterator also takes advantage of this if available at the time
+		// it needs a copy
+		Node[] Children_cachedArray;
+		Node[] Children_GetCachedArray()
+		{
+			if (Children_cachedArray != null) return Children_cachedArray;
+
+			var nodes = new Node[_childCount];
+			var c = _firstChild;
+			int i = 0;
+			while (c != null)
+			{
+				nodes[i++] = c;
+				c = c._nextSibling;
+			}
+			Children_cachedArray = nodes;
+			return nodes;
+		}
+
+		Node Children_ItemAt(int index)
+		{
+			var arr = Children_GetCachedArray();
+			return arr[index];
+		}
+
+		void Children_Invalidate()
+		{
+			if (_safeIterator != null) _safeIterator.SecureCopy();
+			Children_cachedArray = null;
+		}
+
+		IEnumerator<Node> Children_GetEnumerator()
+		{
+			return new SafeIterator(this);
+		}
+
+		SafeIterator _safeIterator; // Linked list
+
+		class SafeIterator: IEnumerator<Node>
+		{
+			Node[] _array;
+			int _pos = -1;
+			Node _current;
+			SafeIterator _nextIterator;
+			Visual _v;
+
+			internal SafeIterator(Visual v)
+			{
+				_v = v;
+				AddToIteratorList();
+			}
+
+			public Node Current
+			{
+				get
+				{
+					if (_array != null) 
+					{
+						if (_array[_pos] == null) throw new Exception();
+						return _array[_pos];
+					}
+					else 
+					{
+						if (_current == null) throw new Exception();
+						return _current;
+					}
+				}
+			}
+
+			public bool MoveNext()
+			{
+				if (_pos == -1)
+				{
+					_array = _v.Children_cachedArray; // If we have a cached array, go ahead and use that
+				}
+
+				_pos++;
+
+				if (_array != null)
+					return _pos < _array.Length;
+				
+				if (_current == null) _current = _v._firstChild;
+				else _current = _current._nextSibling;
+
+				return _current != null;
+			}
+
+			public void Dispose()
+			{
+				Reset();
+				RemoveFromIteratorList();
+				_v = null;
+			}
+
+			public void Reset()
+			{
+				_pos = -1;
+				_current = null;
+				_array = null;
+			}
+
+			void AddToIteratorList()
+			{
+				_nextIterator = _v._safeIterator;
+				_v._safeIterator = this;
+			}
+
+			void RemoveFromIteratorList()
+			{
+				if (_v._safeIterator == this)
+				{
+					_v._safeIterator = _nextIterator;
+				}
+				else
+				{
+					for (var si = _v._safeIterator; si != null; si = si._nextIterator)
+						if (si._nextIterator == this)
+						{
+							si._nextIterator = this._nextIterator;
+							return;
+						}
+				}
+			}
+
+			bool MultipleIterators { get { return _v._safeIterator._nextIterator != null; } }
+
+			internal void SecureCopy()
+			{
+				if (_array == null)
+				{
+					if (MultipleIterators)
```

Comment:
Perhaps check if there is a cachedArray instead, since there might be one even without multiple iterators.